### PR TITLE
Minimize overhead from asynchronous checkpointing

### PR DESCRIPTION
**Pull request by gkroiz**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Minimize overhead as a result of asynchronous checkpointing

**Collection**: [Note which collection this PR will affect]

# Context
I'm from the GPU Goodput & Resiliency team within Google Cloud CMCS. We are benchmarking goodput on nemo 2.0 recipes (in nemo:24.12.01 container) and noticed sub-optimal overhead when enabling asynchronous checkpointing for llama 3 70B and 405B. The observed overhead motivated the two plugins in this PR. These changes are copied from an internal codebase. If there is interest in these changes, we would be happy to clean up and formalize this PR.

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
checkpoint_io = MinCkptOverheadCheckpointIO(
    save_ckpt_format='torch_dist',
    load_directly_on_device=True,
    async_save =args.enable_async_ckpt or args.enable_optimized_async_ckpt,
    torch_dist_multiproc = args.ckpt_threads_per_rank,
    assume_constant_structure = True,
    parallel_save = True,
    parallel_save_within_dp = False,
    parallel_load = True,
)

checkpoint_io = PersistentCheckpointProcessIO(
    checkpoint_io,
    profile_dir = args.trace_dir if args.profile_checkpoint_interval is not None else None,
)

plugins=[checkpoint_io]

trainer = nl.Trainer(
    accelerator="gpu",
    devices=args.num_gpus,
    num_nodes=args.num_nodes,
    max_steps=args.max_steps,
    max_time={"seconds": args.max_runtime},
    callbacks=callbacks,
    log_every_n_steps=None,
    val_check_interval=None, 
    limit_val_batches=None, 
    plugins=plugins,
    strategy=strategy,
    enable_progress_bar=False,
)
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [X] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by akoumpa** [comment]
Hi @gkroiz thanks a lot for the contribution.

To proceed, we have to run CI and to initiate need to pass the copyright-check:

```
Error: Found files with missing copyright:
path= ./nemo/utils/callbacks/persistent_ckpt_proc.py
path= ./nemo/utils/callbacks/min_ckpt_overhead.py
path= ./nemo/utils/callbacks/daemon_strategies.py
path= ./nemo/utils/_ckpt_utils.py
path= ./nemo/utils/_state_dict_utils.py
```

I'd appreciate it if you could add license to these to the top. Thank you.
